File: webapp/php/web/debug/send_web_push.php
#!/usr/bin/env php
<?php
declare(strict_types=1);

use Google\Protobuf\Timestamp;
use Xsuportal\Proto\Resources\Notification;
use Xsuportal\Proto\Resources\Notification\TestMessage;
use App\Application\Handlers\HttpErrorHandler;
use App\Application\Handlers\ShutdownHandler;
use App\Application\ResponseEmitter\ResponseEmitter;
use DI\ContainerBuilder;
use Slim\Factory\AppFactory;
use Slim\Factory\ServerRequestCreatorFactory;

require __DIR__ . '/../vendor/autoload.php';

// start Slim

// Instantiate PHP-DI ContainerBuilder
$containerBuilder = new ContainerBuilder();

if (false) { // Should be set to true in production
	$containerBuilder->enableCompilation(__DIR__ . '/var/cache');
}

// Set up settings
$settings = require __DIR__ . '/app/settings.php';
$settings($containerBuilder);

// Set up dependencies
$dependencies = require __DIR__ . '/app/dependencies.php';
$dependencies($containerBuilder);

// Set up repositories
$repositories = require __DIR__ . '/app/repositories.php';
$repositories($containerBuilder);

// Build PHP-DI Container instance
$container = $containerBuilder->build();

// Instantiate the app
AppFactory::setContainer($container);
$app = AppFactory::create();
$callableResolver = $app->getCallableResolver();

// Register middleware
$middleware = require __DIR__ . '/app/middleware.php';
$middleware($app);

// Register routes
$routes = require __DIR__ . '/app/routes.php';
$routes($app);

/** @var bool $displayErrorDetails */
$displayErrorDetails = $container->get('settings')['displayErrorDetails'];

// Create Request object from globals
$serverRequestCreator = ServerRequestCreatorFactory::create();
$request = $serverRequestCreator->createServerRequestFromGlobals();

// Create Error Handler
$responseFactory = $app->getResponseFactory();
$errorHandler = new HttpErrorHandler($callableResolver, $responseFactory);

// Create Shutdown Handler
$shutdownHandler = new ShutdownHandler($request, $errorHandler, $displayErrorDetails);
register_shutdown_function($shutdownHandler);

// Add Routing Middleware
$app->addRoutingMiddleware();

// Add Error Middleware
$errorMiddleware = $app->addErrorMiddleware($displayErrorDetails, false, false);
$errorMiddleware->setDefaultErrorHandler($errorHandler);

// end Slim


const WEBPUSH_SUBJECT = 'xsuportal-debug@example.com';

function getVpaidKey($path) //
{
    if (file_exists($path)) {
        $privateKey = file_get_contents($path);
        // Webpush::VapidKey.from_pem(private_key)
    } else {
        throw new \UnexpectedValueException('WEBPUSH_VAPID_PRIVATE_KEY_PATH');
    }
}

function makeTestNotificationPb(): Notification
{
    return new Notification([
        'created_at' => new Timestamp(['seconds' => time()]),
        'content_test' => new TestMessage([
            'something' => rand(10000),
        ]),
    ]);
}

function insertNotification(PDO $pdo, Notification $notification, int $contestantId): array
{
    $encodedMessage = base64_encode($notification->serializeTostring());
    $sql = 'INSERT INTO `notifications` (`contestant_id`, `encoded_message`, `read`, `created_at`, `updated_at`) VALUES (?, ?, FALSE, NOW(6), NOW(6))';
    $stmt = $pdo->prepare($sql);
    $stmt->execute([$contestantId, $encodedMessage]);

    $sql = 'SELECT * FROM `notifications` WHERE `id` = LAST_INSERT_ID()';
    $stmt = $pdo->prepare($sql);
    $stmt->execute();

    return $stmt->fetch();
}

function getPushSubscriptions(PDO $pdo, int $contestantId): array
{
    $sql = 'SELECT * FROM `push_subscriptions` WHERE `contestant_id` = ?';
    $stmt = $pdo->prepare($sql);
    $stmt->execute([$contestantId]);

    $subs = $stmt->fetchAll();

    if (count($subs) === 0) {
        throw new \UnexpectedValueException(sprintf('No push subscriptions found: contestant_id=%d', $contestantId));
    }

    return $subs;
}

function sendWebPush($vpaidKey, Notification $notification, $pushSubscription)
{
    $message = base64_encode($notification->serializeToString());
    // vapid = vapid_key.to_h
    $vpaid['subject'] = WEBPUSH_SUBJECT;

    // Webpush.payload_send(
    //     message: message,
    //     endpoint: push_subscription[:endpoint],
    //     p256dh: push_subscription[:p256dh],
    //     auth: push_subscription[:auth],
    //     vapid: vapid,
    // )
}

$contestantId = null;
$vpaidPrivateKeyPath = null;

// contestant_id = nil
// vapid_private_key_path = nil
// option_parser = OptionParser.new do |opt|
//   opt.banner = "Usage: #{__FILE__} -c contestant_id -i vapid_private_key_path"
//   opt.on('-c contestant_id') {|_| contestant_id = _ }
//   opt.on('-i vapid_private_key_path') {|_| vapid_private_key_path = _ }
// end
// option_parser.parse!
//
// abort option_parser.banner if !contestant_id || !vapid_private_key_path

// TODO
$vpaidPrivateKeyPath = $argv[1];
$contestantId = (int)$argv[2];

$vpaidKey = getVpaidKey($vpaidPrivateKeyPath);
$pdo = $app->getContainer()->get(PDO::class);

$subs = getPushSubscriptions($pdo, $contestantId);

$notification = makeTestNotificationPb();
$result = insertNotification($pdo, $notification, $contestantId);
$notification->setId((int)$result['id']);
$notification->setCreatedAt(new Timestamp(['seconds' => strtotime($result['created_at'])]));

echo sprintf('Notification=%s', $notification->serializeToJsonString()) . PHP_EOL;

foreach ($subs as $sub) {
    echo sprintf('Sending web push: push_subscription=%s', $sub) . PHP_EOL;
    sendWebPush($vpaidKey, $notification, $sub);
}

echo 'Finished' . PHP_EOL;

File: webapp/rust/src/bin/send_web_push.rs
use mysql::prelude::*;
use prost::Message;
use rand::prelude::*;
use structopt::StructOpt;

/*
 * これはデバッグ用のコマンドです。使い方は ~isucon/webapp/tools/README.md を見てください。
 */

#[derive(Debug, StructOpt)]
struct Opt {
    #[structopt(short = "c")]
    contestant_id: String,
    #[structopt(short = "i")]
    vapid_private_key_path: String,
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    env_logger::init();

    let opt = Opt::from_args();

    let vapid_key = xsuportal::webpush::VapidKey::open(&opt.vapid_private_key_path)
        .expect("Failed to load VAPID key");

    let mysql_env = xsuportal::MySQLConnectionEnv::default();
    let pool = mysql::Pool::new(
        mysql::OptsBuilder::new()
            .ip_or_hostname(Some(&mysql_env.host))
            .tcp_port(mysql_env.port)
            .user(Some(&mysql_env.user))
            .db_name(Some(&mysql_env.db_name))
            .pass(Some(&mysql_env.password))
            .init(vec!["SET time_zone='+00:00'".to_owned()]),
    )?;
    let mut conn = pool.get_conn()?;
    let subs = get_push_subscriptions(&mut conn, &opt.contestant_id)?;
    let mut notification_pb = make_test_notification_pb();
    let notification = insert_notification(&mut conn, &notification_pb, &opt.contestant_id)?;
    notification_pb.id = notification.id;
    notification_pb.created_at = Some(xsuportal::chrono_timestamp_to_protobuf(
        notification.created_at,
    ));

    println!("Notification={:?}", notification_pb);

    for sub in subs {
        println!("Sending web push: push_subscription={:?}", sub);
        send_web_push(&vapid_key, &notification_pb, &sub).await?;
    }

    println!("Finished");

    Ok(())
}

fn get_push_subscriptions<Q>(
    conn: &mut Q,
    contestant_id: &str,
) -> Result<Vec<xsuportal::PushSubscription>, mysql::Error>
where
    Q: Queryable,
{
    let subs = conn.exec(
        "SELECT * FROM `push_subscriptions` WHERE `contestant_id` = ?",
        (contestant_id,),
    )?;
    if subs.is_empty() {
        panic!(
            "No push subscriptions found: contestant_id={}",
            contestant_id
        );
    } else {
        Ok(subs)
    }
}

fn make_test_notification_pb() -> xsuportal::proto::resources::Notification {
    let now = chrono::Utc::now();
    let mut rng = rand::thread_rng();
    xsuportal::proto::resources::Notification {
        id: 0,
        created_at: Some(xsuportal::chrono_timestamp_to_protobuf(now.naive_utc())),
        content: Some(
            xsuportal::proto::resources::notification::Content::ContentTest(
                xsuportal::proto::resources::notification::TestMessage {
                    something: rng.gen_range(0, 10000),
                },
            ),
        ),
    }
}

fn insert_notification<Q>(
    conn: &mut Q,
    notification_pb: &xsuportal::proto::resources::Notification,
    contestant_id: &str,
) -> Result<xsuportal::Notification, Box<dyn std::error::Error>>
where
    Q: Queryable,
{
    let mut pb = Vec::with_capacity(notification_pb.encoded_len());
    notification_pb.encode(&mut pb)?;
    let encoded_message = base64::encode(&pb);
    conn.exec_drop("INSERT INTO `notifications` (`contestant_id`, `encoded_message`, `read`, `created_at`, `updated_at`) VALUES (?, ?, FALSE, NOW(6), NOW(6))", (contestant_id, encoded_message))?;
    let notification: Option<xsuportal::Notification> =
        conn.query_first("SELECT * FROM `notifications` WHERE `id` = LAST_INSERT_ID()")?;
    Ok(notification.unwrap())
}

async fn send_web_push(
    vapid_key: &xsuportal::webpush::VapidKey,
    notification_pb: &xsuportal::proto::resources::Notification,
    push_subscription: &xsuportal::PushSubscription,
) -> Result<(), Box<dyn std::error::Error>> {
    let mut pb = Vec::with_capacity(notification_pb.encoded_len());
    notification_pb.encode(&mut pb)?;
    let message = base64::encode(&pb);

    let notifier = xsuportal::webpush::WebPushClient {
        endpoint: &push_subscription.endpoint,
        p256dh: &push_subscription.p256dh,
        auth: &push_subscription.auth,
        vapid_key,
        message,
    };

    if let Err(e) = notifier.send().await {
        eprintln!("Web Push service returned an error: {:?}", e);
    }
    Ok(())
}
